docs/man/man3/Sender.3
.TH "Sender" 3 "Version v01.01r" "libcnet" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Sender
.SH SYNOPSIS
.br
.PP
.PP
\fR#include <sender\&.h>\fP
.SS "Data Fields"

.in +1c
.ti -1c
.RI "char * \fB_srcaddress\fP"
.br
.RI "Source Address\&. "
.ti -1c
.RI "struct sockaddr_in \fB_dstaddress\fP"
.br
.RI "The Host receiving the packet\&. "
.ti -1c
.RI "char * \fB_gateway\fP"
.br
.RI "The gateway address of the current Source Address\&. "
.ti -1c
.RI "int \fB_socket\fP"
.br
.RI "File descriptor for the created socket\&. "
.ti -1c
.RI "int \fB_msgcnt\fP"
.br
.RI "Current Message Number\&. "
.ti -1c
.RI "struct protoent * \fB_proto\fP"
.br
.RI "The protocol used\&. "
.ti -1c
.RI "bool \fB_verbose\fP"
.br
.RI "Enable verbosity\&. "
.ti -1c
.RI "struct \fBTimer\fP * \fB_timer\fP"
.br
.RI "A timer synchronized with the receiver\&. "
.ti -1c
.RI "int \fB_mtu\fP"
.br
.RI "The maximum transmission unit\&. "
.ti -1c
.RI "sem_t \fB_mutex\fP"
.br
.RI "A semaphore to synchronize with the \fBReceiver\fP\&. "
.ti -1c
.RI "bool \fB_synch\fP"
.br
.RI "Set to true when it is synchronized with the \fBReceiver\fP\&. "
.ti -1c
.RI "bool \fB_sent\fP"
.br
.RI "Another variable to syncrhonize with the \fBReceiver\fP\&. "
.ti -1c
.RI "struct \fBIpParameters\fP \fB_ipp\fP"
.br
.RI "Some additional parameters for IP Packets\&. "
.ti -1c
.RI "struct \fBIcmpParameters\fP \fB_icmpp\fP"
.br
.RI "Some parameters used for ICMP Packets\&. "
.ti -1c
.RI "struct \fBTcpParameters\fP \fB_tcpp\fP"
.br
.RI "Some parameters used for TCP Packets\&. "
.in -1c
.SH "Detailed Description"
.PP 
Struct representing a \fBSender\fP\&. It is called Raw since it uses the SOCK_RAW option when creating a new socket\&.
.PP
This structure has the following fields:
.PP
.IP "\(bu" 2
\fR_srcaddress\fP The IP address sending the packets
.IP "\(bu" 2
\fR_dstaddress\fP The host receiving the packet
.IP "\(bu" 2
\fR_gateway\fP The gateway address of the current Source address
.IP "\(bu" 2
\fR_socket\fP The file descriptor of the created socket
.IP "\(bu" 2
\fR_msgcnt\fP The current Message number being sent
.IP "\(bu" 2
\fR_proto\fP The protocol used to send the packets
.IP "\(bu" 2
\fR_lastid\fP The last used IP Packet Identifier
.IP "\(bu" 2
\fR_lasticmpid\fP The last used ICMP Packet Identifier
.IP "\(bu" 2
\fR_icmpsn\fP The last ICMP Message Sequence Number
.IP "\(bu" 2
\fR_timer\fP A \fBTimer\fP shared with the \fBReceiver\fP
.IP "\(bu" 2
\fR_mtu\fP The Maximum Transmission Unit
.IP "\(bu" 2
\fR_mutex\fP A simple Mutex to syncrhonize with the receiver
.IP "\(bu" 2
\fR_synch\fP A boolean flag to indicate the syncrhonization
.IP "\(bu" 2
\fR_sent\fP A boolean flag to help avoid deadlock when synchronizing
.IP "\(bu" 2
\fR_params\fP Some additional parameters for IP Packets 
.PP

.PP
Definition at line \fB102\fP of file \fBsender\&.h\fP\&.
.SH "Field Documentation"
.PP 
.SS "struct sockaddr_in _dstaddress"

.PP
The Host receiving the packet\&. 
.PP
Definition at line \fB106\fP of file \fBsender\&.h\fP\&.
.SS "char* _gateway"

.PP
The gateway address of the current Source Address\&. 
.PP
Definition at line \fB107\fP of file \fBsender\&.h\fP\&.
.SS "struct \fBIcmpParameters\fP _icmpp"

.PP
Some parameters used for ICMP Packets\&. 
.PP
Definition at line \fB118\fP of file \fBsender\&.h\fP\&.
.SS "struct \fBIpParameters\fP _ipp"

.PP
Some additional parameters for IP Packets\&. 
.PP
Definition at line \fB117\fP of file \fBsender\&.h\fP\&.
.SS "int _msgcnt"

.PP
Current Message Number\&. 
.PP
Definition at line \fB109\fP of file \fBsender\&.h\fP\&.
.SS "int _mtu"

.PP
The maximum transmission unit\&. 
.PP
Definition at line \fB113\fP of file \fBsender\&.h\fP\&.
.SS "sem_t _mutex"

.PP
A semaphore to synchronize with the \fBReceiver\fP\&. 
.PP
Definition at line \fB114\fP of file \fBsender\&.h\fP\&.
.SS "struct protoent* _proto"

.PP
The protocol used\&. 
.PP
Definition at line \fB110\fP of file \fBsender\&.h\fP\&.
.SS "bool _sent"

.PP
Another variable to syncrhonize with the \fBReceiver\fP\&. 
.PP
Definition at line \fB116\fP of file \fBsender\&.h\fP\&.
.SS "int _socket"

.PP
File descriptor for the created socket\&. 
.PP
Definition at line \fB108\fP of file \fBsender\&.h\fP\&.
.SS "char* _srcaddress"

.PP
Source Address\&. 
.PP
Definition at line \fB105\fP of file \fBsender\&.h\fP\&.
.SS "bool _synch"

.PP
Set to true when it is synchronized with the \fBReceiver\fP\&. 
.PP
Definition at line \fB115\fP of file \fBsender\&.h\fP\&.
.SS "struct \fBTcpParameters\fP _tcpp"

.PP
Some parameters used for TCP Packets\&. 
.PP
Definition at line \fB119\fP of file \fBsender\&.h\fP\&.
.SS "struct \fBTimer\fP* _timer"

.PP
A timer synchronized with the receiver\&. 
.PP
Definition at line \fB112\fP of file \fBsender\&.h\fP\&.
.SS "bool _verbose"

.PP
Enable verbosity\&. 
.PP
Definition at line \fB111\fP of file \fBsender\&.h\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for libcnet from the source code\&.


CHANGELOG.md
# CHANGELOG

## libcnet v01.01r -- 29/06/2024 - 04/07/2024

##### Addition

- Added TCP Header and Packets structure and functions to `ip.h` and `ip.c`
- Added functions for IP Packets to include TCP Packets into `ip.h` and `ip.c`
- Added a version header files containing the current version of the library `utils/version.h`
- Added documentation using Doxygen. Modified all docstrings in according to Doxygen documentation
- Added crafting functions to the Sender for sending TCP Packets, `Sender_craftTcp` and `Sender_fillTcpHeader`
- Added `PseudoHeader` struct to `ip.h` and functions to compute Checksum for ICMP, UDP and TCP

##### Small Changes

- Changed `snprintf` into `sprintf` into `ControlBits_toBin` function of `ip.c`
- Added TCP Option printing informations to the `TcpHeader_printInfo` function

##### Fixed Bugs

- When wrapping a packet into an IP Packet it causes the address of the payload being copied.
  This can cause a double free error when trying to call the `IpPacket_delete` function. To
  avoid this behaviour, first the Header and then the payload are copied into new memory
  locations. Before it was `memcpy(dstpacket, srcpacket, size)` now is
  `memcpy(dstpacket->hdr, srcpacket->hdr, hdr_size)` and `memcpy(dstpacket->payload, srcpacket->payload, psize)`

- When computing the checksum it was jumping two bytes, now it does not. Notice that it computes the checksum
  correctly both for ICMP and UDP Packets, but it does not for TCP Packets and complete IP Packets.

- Correct the bug when creating the TCP Option and then encoding flipping the bytes in an incorrect way.

src/sender.c
#include "sender.h"


Sender* Sender_new(
    const char *_interface, const char* _dstaddr, char* _gateway, const u_int16_t _dstport, 
    const char* _proto, const bool _verbose
) {
    struct sockaddr_in dst;
    int socketfd;

    struct protoent *proto = getprotobyname(_proto);
    socketfd = socket(AF_INET, SOCK_RAW, proto->p_proto);
    if (socketfd == -1) handle_error("socket", socketfd);

    int value = 1;
    setsockopt(socketfd, IPPROTO_IP, IP_HDRINCL, &value, sizeof(value));

    /* Connect to the destination socket */
    memset(&dst, 0, sizeof(dst));
    dst.sin_family = AF_INET;
    dst.sin_port = htons(_dstport);
    inet_pton(AF_INET, _dstaddr, &dst.sin_addr);

    Sender *sender = (Sender*)malloc(sizeof(Sender));

    sender->_srcaddress = (char*)malloc(INET_ADDRSTRLEN * sizeof(char));
    getInterfaceIp(_interface, sender->_srcaddress);

    sender->_dstaddress = dst;
    sender->_gateway = _gateway;
    sender->_socket = socketfd;
    sender->_msgcnt = 0;
    sender->_proto = proto;
    sender->_verbose = _verbose;
    sender->_timer = NULL;
    sender->_mtu = getInterfaceMTU(_interface);
    sender->_synch = false;
    sender->_sent = false;

    // Set default values for Parameters structure
    sender->_ipp._df = D_FLAG_SET;
    sender->_ipp._mf = M_FLAG_NOT_SET;
    sender->_ipp._xf = X_FLAG_NOT_SET;
    sender->_ipp._id = (unsigned short)getpid();
    sender->_icmpp._sn = 0;
    sender->_icmpp._id = 1;
    sender->_tcpp._sn = 1;
    sender->_tcpp._an = 0;
    sender->_tcpp._numOfOptions = 0;

    ControlBits_fromValue(0, &sender->_tcpp._cbits);

    return sender;
}

void Sender_delete(Sender* _self)
{
    if (_self->_synch) sem_destroy(&_self->_mutex);

    TcpOptions_delete(_self->_tcpp._opts, _self->_tcpp._numOfOptions);
    shutdown(_self->_socket, 2);
    free(_self->_srcaddress);
    free(_self);
}

void Sender_setTimer(Sender* _self, struct Timer* _timer)
{
    _self->_timer = _timer;
}

void Sender_setMtu(Sender* _self, const int _mtu)
{
    _self->_mtu	= _mtu;
}

void Sender_setIpFlags(Sender* _self, int _d, int _m)
{
    if (_d == D_FLAG_SET && _m == M_FLAG_SET)
    {
        fprintf(stderr, "[Sender_setIpFlags] Cannot set both D and M flags to 1. ");
        fprintf(stderr, "Program will not terminate but flags will not be set.\n");
    }

    if (
        (_d != D_FLAG_NOT_SET && _d != D_FLAG_SET) &&
        (_m != M_FLAG_NOT_SET && _m != M_FLAG_SET)
    ) {
        fprintf(stderr, "[Sender_setIpFlags] M and D flags must have correct values\n");
        fprintf(stderr, "Program will not terminate but flags will not be set.\n");
    }

    _self->_ipp._xf = X_FLAG_NOT_SET;
    _self->_ipp._df = _d;
    _self->_ipp._mf = _m;
}

void Sender_setTcpOptions(Sender* _self, struct TcpOption** _opts, const int _n)
{
    TcpOptions_copy(_opts, _self->_tcpp._opts, _n);
    _self->_tcpp._numOfOptions = _n;
}

void Sender_bsendto(Sender* _self, const char* _buffer, const size_t _size)
{
    socklen_t dstlen;
    dstlen = sizeof(struct sockaddr_in);
    
    if (
        sendto(
            _self->_socket, _buffer, _size, 0, 
            (struct sockaddr *)&_self->_dstaddress, dstlen
        ) < 0
    ) {
        handle_error("sendto", _self->_socket);
    }

    if (_self->_verbose) printf("[*] Sent %ld bytes of IP Packet\n", _size);
}

void Sender_sendto(Sender* _self, const IpPacket* _pckt)
{
    _self->_sent = false;

    ByteBuffer* buffer = IpPacket_encode(_pckt);

    if (_self->_synch) semaphore_wait(&_self->_mutex, "Sender_sendto");
    
    Sender_bsendto(_self, buffer->_buffer, buffer->_size);

    _self->_msgcnt += 1;

    if (_self->_verbose)
    {
        char filename[100];

        snprintf(
            filename, 60, "sent_%hu_%d_%s.bin", _self->_ipp._id, 
            _self->_msgcnt, _self->_proto->p_name
        );
        
        ByteBuffer_writeToFile(buffer, filename);
    }
    
    ByteBuffer_delete(buffer);

    _self->_sent = true;

    if (_self->_synch) semaphore_post(&_self->_mutex, "Sender_sendto");
}

void Sender_printInfo(const Sender* _self)
{
    u_int16_t dstport = ntohs(_self->_dstaddress.sin_port);

    char dstaddr[INET_ADDRSTRLEN];
    Sender_getDestinationIP(_self, dstaddr);

    printf("[*] Sending To %s:%d\n", dstaddr, dstport);
}

void Sender_getDestinationIP(const Sender* _self, char *_out)
{
    addressNumberToString(_self->_dstaddress.sin_addr.s_addr, _out, true);
}

void Sender_sendc(Sender* _self, const IpPacket* _pckt)
{
    Sender_printInfo(_self);

    while (1)
    {
        Sender_sendto(_self, _pckt);
    }
}

IpPacket* Sender_createIpPacket(Sender *_self, const u_int16_t _id)
{
    u_int8_t  proto   = _self->_proto->p_proto;
    u_int32_t dstaddr = ntohl(_self->_dstaddress.sin_addr.s_addr);
    u_int32_t srcaddr = inet_network(_self->_srcaddress);

    int _m = _self->_ipp._mf;
    int _d = _self->_ipp._df;
    int _x = _self->_ipp._xf;

    IpPacket* ippckt = craftIpPacket(
        IPv4, 0x0, 0x0, IP_HEADER_SIZE, _id, _x, _d, _m, 
        0x0, TTL_DEFAULT_VALUE, proto, 0x0, srcaddr, dstaddr
    );

    return ippckt;
}

void Sender_fillIpHeader(Sender* _self, IpPacket* _pckt)
{
    u_int8_t  proto   = _self->_proto->p_proto;
    u_int32_t dstaddr = ntohl(_self->_dstaddress.sin_addr.s_addr);
    u_int32_t srcaddr = inet_network(_self->_srcaddress);

    int _mf = _self->_ipp._mf;
    int _df = _self->_ipp._df;
    int _xf = _self->_ipp._xf;

    u_int16_t flagoff = computeFlagOff(_xf, _df, _mf, 0);
    u_int16_t dsf = computeDifferentiatedServiceField(0, 0);

    IpPacket_fillHeader(
        _pckt, IPv4, dsf, _pckt->_iphdr._tlength, _self->_ipp._id++, flagoff, 
        TTL_DEFAULT_VALUE, proto, 0, srcaddr, dstaddr
    );
}

void Sender_fillIcmpHeader(
    Sender* _self, IcmpPacket* _pckt, const u_int8_t _type, const u_int8_t _code
) {
    if (
        (
            _type == ICMP_DESTINATION_UNREACHABLE_TYPE ||
            _type == ICMP_SOURCE_QUENCH_TYPE           ||
            _type == ICMP_TIME_EXCEEEDED_TYPE  
        )
    ) {
        if (
            (
                _type == ICMP_DESTINATION_UNREACHABLE_TYPE &&
                _code == ICMP_FRAGMENTATION_NEEDED_CODE
            )
        ) {
            return IcmpPacket_fillHeader_Mtu(_pckt, 0x0, _self->_mtu);
        }

        return IcmpPacket_fillHeader_Unused(_pckt, 0x0);
    }

    if (_type == ICMP_REDIRECT_TYPE)
    {
        return IcmpPacket_fillHeader_Redirect(_pckt, 0x0, inet_network(_self->_gateway));
    }

    if (
        (
            _type == ICMP_ECHO_REPLY_TYPE          ||
            _type == ICMP_ECHO_TYPE                ||
            _type == ICMP_INFORMATION_REQUEST_TYPE ||
            _type == ICMP_INFORMATION_REPLY_TYPE
        )
    ) {
        return IcmpPacket_fillHeader_Echo(_pckt, 0x0, _self->_icmpp._id++, _self->_icmpp._sn++);
    }

    fprintf(stderr, "[Sender_createIcmpPacket] Undefined ICMP type %c\n", _type);
    exit(EXIT_FAILURE);
}

void Sender_fillUdpHeader(Sender* _self, UdpPacket* _pckt, const u_int16_t _srcport) 
{
    u_int16_t dstport = ntohs(_self->_dstaddress.sin_port);
    u_int16_t size = UdpPacket_getPacketSize(_pckt);

    UdpPacket_fillHeader(_pckt, _srcport, dstport, size, 0);
}

void Sender_fillTcpHeader(
    Sender*         _self, TcpPacket*        _pckt, const u_int16_t _port, const u_int8_t _offset, 
    const u_int16_t _window, const u_int16_t _urgpntr
) {
    u_int16_t dstport = ntohs(_self->_dstaddress.sin_port);

    TcpPacket_fillHeader(
        _pckt, _port, dstport, _self->_tcpp._sn, _self->_tcpp._an, _offset,
        _self->_tcpp._cbits, _window, 0, _urgpntr,
        _self->_tcpp._opts, _self->_tcpp._numOfOptions
    );
}

IpPacket* Sender_craftIcmp(
    Sender* _self, const u_int8_t _type, const u_int8_t _code, 
    const char* _payload,  const size_t _size
) {
    IpPacket* pckt = IpPacket_newIcmp(_type, _code, _size);
    Sender_fillIpHeader(_self, pckt);
    Sender_fillIcmpHeader(_self, pckt->_payload._icmp, _type, _code);
    
    if (_payload != NULL && _size > 0)
    {
        IcmpPacket_fillPayload(pckt->_payload._icmp, _payload, _size);
    }

    return pckt;
}

IpPacket* Sender_craftUdp(
    Sender* _self, const u_int16_t _port, const char* _payload, const size_t _size
) {
    IpPacket* pckt = IpPacket_newUdp(_size);
    Sender_fillIpHeader(_self, pckt);
    Sender_fillUdpHeader(_self, pckt->_payload._udp, _port);

    if (_payload != NULL && _size > 0)
    {
        UdpPacket_fillPayload(pckt->_payload._udp, _payload, _size);
    }
    
    return pckt;
}

IpPacket* Sender_craftTcp(
    Sender*         _self,   const u_int16_t _srcport, const u_int8_t _offset, 
    const u_int16_t _window, const u_int16_t _urgpntr, const char*    _payload, 
    const size_t    _size
) {
    IpPacket* pckt = IpPacket_newTcp(_size);
    Sender_fillIpHeader(_self, pckt);
    Sender_fillTcpHeader(_self, pckt->_payload._tcp, _srcport, _offset, _window, _urgpntr);

    if (_payload != NULL && _size > 0)
    {
        TcpPacket_fillPayload(pckt->_payload._tcp, _payload, _size);
    }

    // We need to reset the total length of the IP packet since we may have
    // added some TCP options when filling the IP header
    IpHeader_setTotalLength(
        &pckt->_iphdr, IP_HEADER_SIZE + TcpPacket_getPacketSize(pckt->_payload._tcp)
    );

    return pckt;
}

void Sender_send(Sender* _self, IpPacket* _pckt, const double _delay)
{
    if (_self->_timer != NULL) Timer_resetPrevious(_self->_timer);

    // Before sending, let's compute the checksum
    IpPacket_computeChecksum(_pckt);

    // Then send the packet
    Sender_sendto(_self, _pckt);
    
    // Sleep using the created timer
    Timer_sleep(_delay);
}

void Sender_updateIcmpPacket(Sender* _self, IpPacket* _pckt)
{
    IpHeader_setIdentfication(&_pckt->_iphdr, _self->_ipp._id++);
    
    u_int8_t _type = _pckt->_payload._icmp->_hdr._type;

    if (
        (
            _type == ICMP_ECHO_TYPE                ||
            _type == ICMP_INFORMATION_REQUEST_TYPE
        )
    ) {
        IcmpHeader_setIdentifier(&_pckt->_payload._icmp->_hdr, _self->_icmpp._id++);
        IcmpHeader_setSequenceNumber(&_pckt->_payload._icmp->_hdr, _self->_icmpp._sn++);
    }
}

void Sender_updateUdpPacket(Sender* _self, IpPacket* _pckt)
{
    IpHeader_setIdentfication(&_pckt->_iphdr, _self->_ipp._id++);
}